README.md
# Nil-ng

A modern port of the class game Nil, which is a Liero clone.


ChangeLog
Changelog for NiL

	-----------------------------------------------------------
Notice about NiL Changelog usage and formatting.
Please use ONLY the following formatting in this Changelog:
 @DATE @PERSON
 	* @CHANGE

	@DATE is date of commiting to CVS (changes that are not commited to
	CVS should NEVER be added to Changelog). Local time is ok.

	@PERSON is the full REAL name of the one commiting to CVS (not necceserely
	the code author). If you wish to add your nick/handle, add it after your
	name in parenthesis.

	@CHANGE is a short description of the modification done. If you are not
	the author of parts/all of the code, add note about the original author
	in parenthesis.

ALWAYS keep line lenghts UNDER/AT 80 characters.

	-----------------------------------------------------------

16/05/2005 Christoph Brill (egore
	* gcc4 warning fixes
	* less output for mouse and warning fix
	* segfault fix for 0 joysticks

24/03/2005 Harri Ilari Tapio Liusvaara (ilari)
	* make scons work
	* enhance vfs

24/03/2005 Christoph Brill (egore)
	* add checks in scons for zlib and some SDL libs

06/03/2005 Harri Ilari Tapio Liusvaara (hliusvaa)
	* major rework of the vfs

16/01/2005 Christoph Brill (egore)
	* make use of get_text_geom()'s return value

15/01/2005 Christoph Brill (egore)
	* make use of event_target's methods

13/01/2005 Harri Ilari Tapio Liusvaara (hliusvaa)
	* vfs update (changes for pathexpander)

13/01/2005 Christoph Brill (egore)
	* joystick support
	* mouse support

12/01/2005 Christoph Brill (egore)
	* move keymapper to client
	* move Strless to misc.h
	* compile fix for object.h
	* make use of events properly
	* direcotry move src/common/controls to src/client/controls
	* miscelanious fixes I forgot

06/01/2005 Christoph Brill (egore)
	* centered menu

05/01/2005 Christoph Brill (egore)
	* directory move

04/01/2005 Christoph Brill (egore)
	* Code and API cleanup
	* Begin of direcotry move (starting with net)

03/01/2005 Christoph Brill (egore)
	* API and code cleanup

14/12/2004 Christoph Brill (egore)
	* GCC 3.4 fix for joysticks

14/12/2004 Harri Ilari Tapio Liusvaara (hliusvaa)
	* re-keymapping for all keys now allows
	* configure script
	 * fixed circular dependecy problem
	 * changed from objdump to ld

21/11/2004 Harri Ilari Tapio Liusvaara (hliusvaa)
	* add CPIO maker
	* update rgbf<->png converter

20/11/2004 Harri Ilari Tapio Liusvaara (hliusvaa)
	* fix gzip/bzip2 and lump-loading double-free vulnerabilities

18/11/2004 Christoph Brill (egore)
	* replace Makefiles by configure files for whole NiL

18/11/2004 Harri Ilari Tapio Liusvaara (hliusvaa)
	* add new graphics converters
	 * png2rgbf
	 * rgbf2png

11/11/2004 Harri Ilari Tapio Liusvaara (hliusvaa)
	* new console wiring
	 * nilConsole slave
	 * stdout slave

11/11/2004 Christoph Brill (egore)
	* removed World's isPassable, and renamed isPassableCheck to
	  isPassable

11/09/2004 Christoph Brill (egore)
	* Doxygen updated
	 * @param and @return introduction
	 * some descriptions added

11/06/2004 Christoph Brill (egore)
	* configure autodetection fix
	* Doxygen updates
	* new constructor in fonts.h
	* minor API cleanups

10/23/2004 Christoph Brill (egore)
	* added submenus

10/23/2004 Christoph Brill (egore)
	* enhanced menu code
	* fix console buffer overflow

10/23/2004 Harri Ilari Tapio Liusvaara (hliusvaa)
	* add new console base (not used yet)

10/23/2004 Christoph Brill (egore)
	* fixed font loading code (thanks to ilari)

10/23/2004 Christoph Brill (egore)
	* add missing entry to ./congigure --help
	* make NiL compile with -Wall (one issue remaining)
	* add -Wall to default ./configure

10/22/2004 Christoph Brill (egore)
	* first steps towards a menu

09/25/2004 Christoph Brill (egore)
	* patches from TPC on #nil
	 * gcc 3.4 fixes
	* configure fix

09/24/2004 Harri Ilari Tapio Liusvaara (hliusvaa)
	* moved logging from fluff.cpp to logmsg.cpp
	* change logging funktion to iostream
	* enhanced configure for building cpio file
	* include SDL-headers via systemheaders.h
	* add suport for builtin lumps
	* limit cpio memory allocation

09/23/2004 Christoph Brill (egore)
	* enhanced configure for --enable-profiling

09/09/2004 Christoph Brill (egore)
	* adjusted plasmagun

09/08/2004 Christoph Brill (egore)
	* fixed randomizer

09/08/2004 Harri Ilari Tapio Liusvaara (hliusvaa)
	* fixed lump creation
	 * allocate only memory we need
	 * open the file in binary mode

09/08/2004 Christoph Brill (egore)
	* added sound for the railgun

09/08/2004 Phil Howlett (phowlett)
	* added double buffering support
	  This should remove the screen flickering under video displays
	  that are using Hardware acceleration. Quite noticable under
	  win32 machines with fancy video cards.

09/07/2004 Christoph Brill (egore)
	* enhanced Inno Setup Script
	 * german localization
	 * spelling fixes
	 * added shortcur for bot
	 * cut into components main, data and docs
	 * added URL

09/07/2004 Phil Howlett (phowlett)
	* added Inno Setup Script for win32 installation
	* added documentation for win32 compilation
	* added NiL icon (quite ugly)
	* added default NiL configuration file

09/07/2004 Christoph Brill (egore)
	* use SDL_GetError() in view_rgbf.cpp

09/04/2004 Christoph Brill (egore)
	* fix different uninitialized values

09/03/2004 Christoph Brill (egore)
	* reintroduced weapons
	 * all expect railgun
	 * fixed owner id issues
	* tried fixing collision detection

08/26/2004 Harri Ilari Tapio Liusvaara (hliusvaa)
	* Attached is patch to rewrite NiL VFS.
	* New features
	 * Use new SVR4 without CRC CPIO archive format as lumps
	   (GNU CPIO can now pack and unpack lump files)
	 * GZIP/BZIP2 compressed lump decompression on load
	   (new code can load lumps compressed with GZIP or BZIP2)
	 * Use Mmap always if possible
	* Attached is patch to rewrite NiL VFS.
	* New features
	 * Use new SVR4 without CRC CPIO archive format as lumps
	   (GNU CPIO can now pack and unpack lump files)
	 * GZIP/BZIP2 compressed lump decompression on load

src/nil_main.cpp
/***************************************************************************
  nil_main.cpp  -  NEEDDESCRIPTION
 ---------------------------------------------------------------------------
  begin                : 1999-11-15
  by                   : Flemming Frandsen
  email                : dion@swamp.dk

  last changed         : 2004-03-28
  by                   : Christoph Brill
  email                : egore@gmx.de

  last changed         : 2004-09-18
  by                   : H. Ilari Liusvaara
  email                : hliusvaa@cc.hut.fi
  changedescription    : added logmsg.h header.

  last changed         : 2004-09-22
  by                   : H. Ilari Liusvaara
  email                : hliusvaa@cc.hut.fi
  changedescription    : Use load_files_standard().

  last changed         : 2004-10-23
  by                   : H. Ilari Liusvaara
  email                : hliusvaa@cc.hut.fi
  changedescription    : Make simple to stdout console and wire it to master.

  last changed         : 2004-10-28
  by                   : H. Ilari Liusvaara
  email                : hliusvaa@cc.hut.fi
  changedescription    : Use Console::set_source instead of setmsgsource.
                         Use Console::set_chattiness instead of setchattiness.
                         Use Console::put_output instead of fprintf.

  last changed         : 2004-01-14
  by                   : H. Ilari Liusvaara
  email                : hliusvaa@cc.hut.fi
  changedescription    : Fix it so it compiles after vfs_pathexpander.h
                         changes.

  copyright            : (C) 1999 by Flemming Frandsen
  email                : dion@swamp.dk

 ***************************************************************************/

#include "client/graphics/raw_surface.h"
#include "client/graphics/view_rgbf.h"
#include "client/nil_client.h"
#include "client/bot.h"
#include "common/console/logmsg.h"
#include "common/console/console.h"
#include "common/configuration/configuration.h"
#include "common/misc.h"
#include "common/systemheaders.h"
#include "common/vfs/loader.h"
#include "server/nil_server.h"
#include <csignal>
#include <vector>

#ifdef _WIN32
int win32_main(int argc, char *argv[]);
#define NILRC "nil.cfg"
#else
#define NILRC "~/.nilrc"
#endif
#define NILRC_LAST NILRC ".last_good"

char szNILRC[] = NILRC;
char szNILRC_LAST[] = NILRC_LAST;

//! @author H. Ilari Liusvaara
//! @brief Simple print-to-stdout console.
class Console_stdout : public Console
{
	/*! ???
	    @param severity ???
	    @param source ???
	    @param output ??? */
	void _put_output(Console::Severity severity, const char* source, const char* output) throw() {
	//No need to mask control characters as Console::put_output does it automatically.
#ifdef DONT_USE_COLORS
		const char* severities[] = {"Debug", "Message", "Warning", "Error", "Fatal", "Everything's fucked up", 
			"Assertion violation" };
		printf("%s: %s: %s", source, severities[severity], output);
#else
		const char* severities[] = {"\033[1;32m", "\033[1;37m", "\033[1;33m", "\033[1;31m", "\033[0;31m", "\033[1;35m",
			"\033[1;34m" };
		printf("%s%s: %s\033[0m", severities[severity], source, output);
#endif
	}
};


static void signal_handler(int sig)
{
	raise(SIGSEGV);
}

//! the "real" nil main function
int nil_main(int argc,char **argv) {
	// Wire our stdout console to master.
	Console_stdout stdout_console;
	master_console.add_slave(stdout_console);

	signal(SIGABRT, signal_handler);

	// The string that appears before the log messages
	master_console.set_source("NiL");

	// check a few things:
	if (strcmp(OBJECT_NAMES[ot_error],"ERROR")) {
		// error object not last in array -> needed for return values?
		logmsg(lt_fatal,"You have fucked up, OBJECT_NAMES[ot_error] was not ERROR, it was: \"%s\" take a look at fluff.h",OBJECT_NAMES[ot_error]);
		return 1;
	}

	master_console.put_output(Console::MESSAGE, "NiL - Worms, Violence and the GPL\n\n");
	Configuration config;
	
	if (!config.load_file(szNILRC)) {
		master_console.put_output(Console::WARNING, "Unable to read " NILRC ":\n%s\n",config.get_error());
	}

	if (!config.parse_cmdline(argc,argv)) {
		master_console.put_output(Console::FATAL, "Unable to understand commandline: %s\n",config.get_error());
		master_console.put_output(Console::FATAL, "\nUsage: nil [<option>=<value>]\n");
		master_console.put_output(Console::FATAL, "Options can be given in " NILRC " configuration file or\n");
		master_console.put_output(Console::FATAL, "overridden from the command line ('nil mode=server').\n\n");
		master_console.put_output(Console::FATAL, "The options I understand (on commandline or in ~/.nilrc) are:\n");
		config.print_manual(stdout);
		return 1;
	}

	if (!config.check_options()) {
		master_console.put_output(Console::FATAL, "A value failed the secondary check:\n%s\n",config.get_error());
		return 2;
	}

	config.save_options(szNILRC_LAST);

    char szChattiness[] = "chattiness";
	const char* chattiness = config.get_option(szChattiness);
	Console::Severity chattiness_numeric = Console::MESSAGE;
	if(!strcasecmp(chattiness, "debug") || !strcasecmp(chattiness, "4"))
		chattiness_numeric = Console::DEBUG;
	else if(!strcasecmp(chattiness, "message") || !strcasecmp(chattiness, "3"))
		chattiness_numeric = Console::MESSAGE;
	else if(!strcasecmp(chattiness, "warning") || !strcasecmp(chattiness, "2"))
		chattiness_numeric = Console::WARNING;
	else if(!strcasecmp(chattiness, "error") || !strcasecmp(chattiness, "1"))
		chattiness_numeric = Console::ERROR;
	else if(!strcasecmp(chattiness, "fatal") || !strcasecmp(chattiness, "0"))
		chattiness_numeric = Console::FATAL;

	master_console.set_chattiness(chattiness_numeric);

    char szMode[] = "mode";
	const char *mode = config.get_option(szMode);

	if (strcasecmp(mode,"server") == 0) {
		run_server(config);
	} else if (strcasecmp(mode,"client") == 0) {
		run_client(config);
	} else if (strcasecmp(mode,"bot") == 0) {
		run_bot(config);
	} else if (strcasecmp(mode,"serverclient") == 0) {
#ifdef _WIN32
		std::vector<const char*> args(argv,argv+argc);
		args.push_back("mode=client");
		args.push_back(NULL);

		int pi = _spawnv(_P_NOWAIT, argv[0], (char * const *)&args[0]);
		run_server(config);
		TerminateProcess((HANDLE)pi, 0);
#else
		logmsg(lt_message,"Going to fork off a client and run a server...");
		//fork off a client before running the server
		pid_t pid = fork();
		if (pid == 0)
			run_client(config);
		else {
			run_server(config);
			kill(pid,SIGTERM);
		}
#endif
	} else if (strcasecmp(mode,"clientserver") == 0) {
#ifdef _WIN32
		std::vector<const char*> args(argv,argv+argc);
		args.push_back("mode=server");
		args.push_back(NULL);

		int pi = _spawnv(_P_NOWAIT, argv[0], (char * const *)&args[0]);
		run_client(config);
		TerminateProcess((HANDLE)pi, 0);
#else
		logmsg(lt_message,"Going to fork off a server and run a client...");
		//fork off a server before running the client
		pid_t pid = fork();
		if (pid == 0)
			run_server(config);
	 	else {
			run_client(config);
			kill(pid,SIGTERM);
		}
#endif
	} else if (strcasecmp(mode,"cleanup") == 0) {
		logmsg(lt_debug,"This mode is obsolete");
	} else if (strcasecmp(mode,"savelump") == 0) {

		//expand the output filename:
		char expanded_filename[200];
        char szOutputLump[] = "output_lump";
		try {
			VFS::expand_path(expanded_filename, 
				sizeof(expanded_filename),
				config.get_option(szOutputLump));
		} catch(std::exception& e) {
			cerror << "Unable to expand the path \"" 
				<< config.get_option(szOutputLump) << "\": "
				<< e.what() << "\n";
			return false;
		}

		Loader loader;
        char szData[] = "data";
		load_files_standard(loader, config.get_option(szData));

		//save a lump:
		int bs = loader.save_lump(NULL, 0);
		char *buffer = (char*)malloc(bs);
		int lump_size= loader.save_lump(buffer, bs);
		if (lump_size > 0) {
			logmsg(lt_debug, "Saving %i bytes to %s", lump_size, expanded_filename);
			FILE *file = fopen(expanded_filename, "wb");
			fwrite(buffer, lump_size, 1, file);
			fclose(file);
		}
		free(buffer);
	} else if(strcasecmp(mode,"viewfile")==0) {
		view_rgbf(config);	  
	} else {
		logmsg(lt_fatal,"This cannot possibly happen, mode was: '%s'",mode);
		return 3;
	}
	return 0;
}

#ifdef _WIN32
//! main function to determine what to do...
extern "C" int SDL_main(int argc, char *argv[]) {
	// prepare win32 stuff (also calls nil_main)...
	return win32_main(argc, argv);
}
#else
//! main function to determine what to do...
int main(int argc, char **argv) {
	// unix, simply start
	return nil_main(argc, argv);
}
#endif
